drivers/rtc/rtc_stm32.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2020 Paratronic
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Source file for the STM32 RTC driver
+ *
+ */
+
+
+#include <kernel.h>
+#include <soc.h>
+#include <drivers/clock_control/stm32_clock_control.h>
+#include <drivers/rtc.h>
+#include <sys/timeutil.h>
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(rtc, LOG_LEVEL_DBG);
+
+#if defined(CONFIG_RTC_STM32_CLOCK_LSI)
+/* prescaler values for LSI @ 32 KHz */
+#define RTC_PREDIV_ASYNC 31
+#define RTC_PREDIV_SYNC  999
+#else /* CONFIG_COUNTER_RTC_STM32_CLOCK_LSE */
+/* prescaler values for LSI @ 32768 Hz */
+#define RTC_PREDIV_ASYNC 31
+#define RTC_PREDIV_SYNC  1023
+#endif
+
+static int rtc_stm32_settime(struct device *dev, const struct timespec *tp)
+{
+	int err = -EIO;
+	struct tm tm;
+	LL_RTC_TimeTypeDef timeDef;
+	LL_RTC_DateTypeDef dateDef;
+	const char *serr = NULL;
+
+	ARG_UNUSED(dev);
+
+	k_sched_lock();
+
+	LL_PWR_EnableBkUpAccess();
+
+	gmtime_r(&tp->tv_sec, &tm);
+	timeDef.Hours = tm.tm_hour;
+	timeDef.Minutes = tm.tm_min;
+	timeDef.Seconds = tm.tm_sec;
+	if (LL_RTC_TIME_Init(RTC, LL_RTC_FORMAT_BIN, &timeDef) != SUCCESS) {
+		serr = "failed to set time";
+		goto out;
+	}
+
+	dateDef.WeekDay = ((tm.tm_wday + 6) % 7) + 1;
+	dateDef.Month = tm.tm_mon + 1;
+	dateDef.Day = tm.tm_mday;
+	dateDef.Year = (tm.tm_year + 1900) - 2000;
+	if (LL_RTC_DATE_Init(RTC, LL_RTC_FORMAT_BIN, &dateDef) != SUCCESS) {
+		serr = "failed to set date";
+		goto out;
+	}
+
+	err = 0;
+out:
+	LL_PWR_DisableBkUpAccess();
+
+	k_sched_unlock();
+
+	if (serr) {
+		LOG_ERR("%s", serr);
+	}
+
+	return err;
+}
+
+static int rtc_stm32_gettime(struct device *dev, struct timespec *tp)
+{
+	struct tm tm;
+	u32_t time;
+	u32_t subSeconds;
+	u32_t date;
+
+	ARG_UNUSED(dev);
+
+	subSeconds = LL_RTC_TIME_GetSubSecond(RTC);
+	time = LL_RTC_TIME_Get(RTC);
+	date = LL_RTC_DATE_Get(RTC);
+
+	tm.tm_sec = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_SECOND(time));
+	tm.tm_min = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_MINUTE(time));
+	tm.tm_hour = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_HOUR(time));
+
+	tm.tm_mday = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_DAY(date));
+	tm.tm_mon = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_MONTH(date)) - 1;
+	tm.tm_year = __LL_RTC_CONVERT_BCD2BIN(__LL_RTC_GET_YEAR(date)) + 100;
+	tp->tv_sec = timeutil_timegm(&tm);
+	tp->tv_nsec = ((RTC_PREDIV_SYNC - subSeconds) * 1000000000) /
+		(RTC_PREDIV_SYNC + 1);
+
+	return 0;
+}
+
+static int rtc_stm32_init(struct device *dev)
+{
+	int err = -EIO;
+	struct device *clk;
+	struct stm32_pclken pclken;
+	LL_RTC_InitTypeDef init;
+
+	clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	__ASSERT_NO_MSG(clk);
+
+	pclken.enr = DT_INST_0_ST_STM32_RTC_CLOCK_BITS;
+	pclken.bus = DT_INST_0_ST_STM32_RTC_CLOCK_BUS;
+	clock_control_on(clk, (clock_control_subsys_t *) &pclken);
+
+	LL_PWR_EnableBkUpAccess();
+
+#if defined(CONFIG_RTC_STM32_CLOCK_LSI)
+
+#if defined(CONFIG_SOC_SERIES_STM32WBX)
```

Comment:
Please remove any code specific to series that are not L4.

Reply:
It's done!